plugins/events-backend/src/service/hub/MemoryEventBusStore.ts

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright 2024 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { EventParams } from '@backstage/plugin-events-node';
+import { EventBusStore } from './types';
+import { NotFoundError } from '@backstage/errors';
+
+const MAX_BATCH_SIZE = 10;
+
+export class MemoryEventBusStore implements EventBusStore {
+  #events = new Array<
+    EventParams & { seq: number; notifiedSubscribers: Set<string> }
+  >();
+  #subscribers = new Map<
+    string,
+    { id: string; seq: number; topics: Set<string> }
+  >();
+  #listeners = new Set<{
+    topics: Set<string>;
+    resolve(result: { topic: string }): void;
+  }>();
+
+  async publish(options: {
+    event: EventParams;
+    notifiedSubscribers: string[];
+  }): Promise<{ eventId: string } | undefined> {
+    const topic = options.event.topic;
+    const notifiedSubscribers = new Set(options.notifiedSubscribers);
+
+    let hasOtherSubscribers = false;
+    for (const sub of this.#subscribers.values()) {
+      if (sub.topics.has(topic) && !notifiedSubscribers.has(sub.id)) {
+        hasOtherSubscribers = true;
+        break;
+      }
+    }
+    if (!hasOtherSubscribers) {
+      return undefined;
+    }
+
+    const nextSeq = this.#getMaxSeq() + 1;
+    this.#events.push({ ...options.event, notifiedSubscribers, seq: nextSeq });
+
+    for (const listener of this.#listeners) {
+      if (listener.topics.has(topic)) {
+        listener.resolve({ topic });
+        this.#listeners.delete(listener);
+      }
+    }
+    return { eventId: String(nextSeq) };
+  }
+
+  #getMaxSeq() {
+    return this.#events[this.#events.length - 1]?.seq ?? 0;
+  }
+
+  async upsertSubscription(id: string, topics: string[]): Promise<void> {
+    const existing = this.#subscribers.get(id);
+    if (existing) {
+      existing.topics = new Set(topics);
+      return;
+    }
+    const sub = {
+      id: id,
+      seq: this.#getMaxSeq(),
+      topics: new Set(topics),
+    };
+    this.#subscribers.set(id, sub);
+  }
+
+  async readSubscription(id: string): Promise<{ events: EventParams[] }> {
+    const sub = this.#subscribers.get(id);
+    if (!sub) {
+      throw new NotFoundError(`Subscription not found`);
+    }
+    const events = this.#events
+      .filter(
+        event =>
+          event.seq > sub.seq &&
```

Comment:
Skipped it to keep it simple since this isn't intended to be used in production. It does mean we might want to take a different approach to MySQL if sticking to that though. I'll explore cleanup here or different handling in MySQL.

One potential idea is to simply have the MySQL implementation just throw away all incoming events, it's almost the same thing since we forward events locally too. Ofc if you only split deploy without horizontal scaling you benefit from the memory story, but a bit strange to only scale that way anyway tbh.

Reply:
Ended up trimming old events, most of the implementation already supports it